File: CoreSamples/Bit6FullDemo-ObjC/ChatDemo/ChatsTableViewController.m
//
//  ChatsViewController.m
//  Bit6ChatDemo
//
//  Created by Carlos Thurber Boaventura on 04/01/14.
//  Copyright (c) 2014 Bit6. All rights reserved.
//

#import "ChatsTableViewController.h"
#import <MobileCoreServices/MobileCoreServices.h>
#import "ConversationsViewController.h"

@interface ChatsTableViewController () <UITextFieldDelegate>
{
    BOOL scroll;
}

@property (strong, nonatomic) NSMutableArray *messages;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *typingBarButtonItem;

@end

@implementation ChatsTableViewController

- (void)setConversation:(Bit6Conversation *)conversation
{
    _conversation = conversation;
    [Bit6 setCurrentConversation:conversation];
    if (conversation.address.isGroup) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(groupsChangedNotification:) name:Bit6GroupsChangedNotification object:nil];
    }
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(messagesChangedNotification:) name:Bit6MessagesChangedNotification object:self.conversation];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(typingBeginNotification:) name:Bit6TypingDidBeginRtNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(typingEndNotification:) name:Bit6TypingDidEndRtNotification object:nil];
}

- (NSArray*)messages
{
    if (!_messages && self.conversation) {
        _messages = [NSMutableArray arrayWithArray:self.conversation.messages];
    }
    return _messages;
}

#pragma mark -

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationItem.prompt = [NSString stringWithFormat:@"Logged as %@",Bit6.session.activeIdentity.uri];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"bit6ui_phone"] style:UIBarButtonItemStylePlain target:self action:@selector(call)];
    
    if (self.conversation.typingAddress) {
        self.typingBarButtonItem.title = [NSString stringWithFormat:@"%@ is typing...",self.conversation.typingAddress.uri];
    }
    else {
        self.typingBarButtonItem.title = nil;
    }
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.navigationController setToolbarHidden:NO animated:YES];
    if (!scroll) {
        scroll = YES;
        [self.tableView reloadData];
        [self.tableView setContentOffset:CGPointMake(0, CGFLOAT_MAX) animated:NO];
    }
}

- (void)dealloc
{
    [Bit6 setCurrentConversation:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.messages.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    Bit6Message *message = self.messages[indexPath.row];
    
    static NSString *textInCellIdentifier = @"textInCell";
    static NSString *textOutCellIdentifier = @"textOutCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:message.incoming?textInCellIdentifier:textOutCellIdentifier];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    Bit6Message *message = self.messages[indexPath.row];
    
    UILabel *textLabel = (UILabel*) [cell viewWithTag:1];
    textLabel.text = [ChatsTableViewController textContentForMessage:message];
    
    UILabel *detailTextLabel = (UILabel*) [cell viewWithTag:2];
    if (detailTextLabel) {
        if (message.type == Bit6MessageType_Call) {
            detailTextLabel.text = @"";
        }
        else {
            switch (message.status) {
                case Bit6MessageStatus_New : detailTextLabel.text = @""; break;
                case Bit6MessageStatus_Sending : detailTextLabel.text = @"Sending"; break;
                case Bit6MessageStatus_Sent : detailTextLabel.text = @"Sent"; break;
                case Bit6MessageStatus_Failed : detailTextLabel.text = @"Failed"; break;
                case Bit6MessageStatus_Delivered : detailTextLabel.text = @"Delivered"; break;
                case Bit6MessageStatus_Read : detailTextLabel.text = @"Read"; break;
            }
        }
    }
    
}

#pragma mark - Calls

- (void)call {
    UIAlertController* actionSheet = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"Audio Call" style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * action) {
                                                      [Bit6 startCallTo:self.conversation.address streams:Bit6CallStreams_Audio];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"Video Call" style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * action) {
                                                      [Bit6 startCallTo:self.conversation.address streams:Bit6CallStreams_Audio|Bit6CallStreams_Video];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    
    [self.navigationController presentViewController:actionSheet animated:YES completion:nil];
}

#pragma mark - Send Text

- (IBAction)touchedComposeButton:(id)sender {
    if (!self.canChat) {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"You have left this group" message:nil preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
        [self.navigationController presentViewController:alert animated:YES completion:nil];
        return;
    }
    
    UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Type the message" message:nil preferredStyle:UIAlertControllerStyleAlert];
    __weak typeof(self) weakSelf = self;
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.delegate = weakSelf;
    }];
    [alert addAction:[UIAlertAction actionWithTitle:@"Send" style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * action) {
                                                      UITextField *textField = alert.textFields[0];
                                                      NSString *msg = textField.text;
                                                      
                                                      Bit6OutgoingMessage *message = [Bit6OutgoingMessage new];
                                                      message.content = msg;
                                                      message.destination = self.conversation.address;
                                                      [message sendWithCompletionHandler:^(NSDictionary *response, NSError *error) {
                                                          if (!error) {
                                                              NSLog(@"Message Sent");
                                                          }
                                                          else {
                                                              NSLog(@"Message Failed with Error: %@",error.localizedDescription);
                                                          }
                                                      }];
                                                  }]];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    
    [self.navigationController presentViewController:alert animated:YES completion:nil];
}

#pragma mark - Typing

- (void)typingBeginNotification:(NSNotification*)notification
{
    Bit6Address* fromAddress = notification.userInfo[Bit6FromKey];
    Bit6Address* convervationAddress = notification.object;
    if ([convervationAddress isEqual:self.conversation.address]) {
        self.typingBarButtonItem.title = [NSString stringWithFormat:@"%@ is typing...",fromAddress.uri];
    }
}
- (void)typingEndNotification:(NSNotification*)notification
{
    Bit6Address* convervationAddress = notification.object;
    if ([convervationAddress isEqual:self.conversation.address]) {
        self.typingBarButtonItem.title = nil;
    }
}

#pragma mark - Data Source changes

//here we listen to changes in group members and group title
- (void)groupsChangedNotification:(NSNotification*)notification
{
    Bit6Group *object = notification.userInfo[Bit6ObjectKey];
    NSString *change = notification.userInfo[Bit6ChangeKey];
    
    if ([change isEqualToString:Bit6UpdatedKey] && [object.address isEqual:self.conversation.address]) {
        self.title = [ConversationsViewController titleForConversation:self.conversation];
    }
}

- (void)messagesChangedNotification:(NSNotification*)notification
{
    Bit6Message *object = notification.userInfo[Bit6ObjectKey];
    NSString *change = notification.userInfo[Bit6ChangeKey];
    
    if ([change isEqualToString:Bit6AddedKey]) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:self.messages.count inSection:0];
        [self.messages addObject:object];
        [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        [self scrollToBottomAnimated:YES];
    }
    else if ([change isEqualToString:Bit6UpdatedKey]) {
        NSIndexPath *indexPath = [self findMessage:object];
        if (indexPath) {
            UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
            [self tableView:self.tableView willDisplayCell:cell forRowAtIndexPath:indexPath];
        }
    }
    else if ([change isEqualToString:Bit6DeletedKey]) {
        NSIndexPath *indexPath = [self findMessage:object];
        if (indexPath) {
            [self.messages removeObjectAtIndex:indexPath.row];
            [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        }
    }
}

- (NSIndexPath*)findMessage:(Bit6Message*)message
{
    NSUInteger index = NSNotFound;
    for (NSInteger x=self.messages.count-1; x>=0; x--) {
        if ([self.messages[x] isEqual:message]) {
            index = x;
            break;
        }
    }
    return [NSIndexPath indexPathForRow:index inSection:0];
}

#pragma mark -

- (void)scrollToBottomAnimated:(BOOL)animated
{
    if (self.messages.count>0) {
        long section = 0;
        long row = [self tableView:self.tableView numberOfRowsInSection:section] - 1;
        NSIndexPath *scrollIndexPath = [NSIndexPath indexPathForRow:row inSection:0];
        [self.tableView scrollToRowAtIndexPath:scrollIndexPath atScrollPosition:UITableViewScrollPositionBottom animated:animated];
    }
}

#pragma mark - UITextFieldDelegate

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    [Bit6 typingBeginToAddress:self.conversation.address];
    return YES;
}

#pragma mark - Helpers

- (BOOL)canChat
{
    if (self.conversation.address.isGroup) {
        Bit6Group *group = [Bit6Group groupWithConversation:self.conversation];
        if (group.hasLeft) {
            return NO;
        }
    }
    return YES;
}

+ (NSString*)textContentForMessage:(nonnull Bit6Message*)message
{
    if (message.type == Bit6MessageType_Call) {
        
        BOOL showDuration = NO;
        
        NSString *status = @"";
        switch (message.callStatus) {
            case Bit6MessageCallStatus_Answer: status = @"Answer"; showDuration = YES; break;
            case Bit6MessageCallStatus_Missed: status = @"Missed"; break;
            case Bit6MessageCallStatus_Failed: status = @"Failed"; break;
            case Bit6MessageCallStatus_NoAnswer: status = @"No Answer"; break;
        }
        
        NSMutableArray *channels = [NSMutableArray arrayWithCapacity:3];
        if ([message callHasChannel:Bit6MessageCallChannel_Audio]){
            [channels addObject:@"Audio"];
        }
        if ([message callHasChannel:Bit6MessageCallChannel_Video]){
            [channels addObject:@"Video"];
        }
        if ([message callHasChannel:Bit6MessageCallChannel_Data]){
            [channels addObject:@"Data"];
        }
        
        NSString *channel = [channels componentsJoinedByString:@" + "];
        
        if (showDuration) {
            return [NSString stringWithFormat:@"%@ Call - %@s",channel, message.callDuration.description];
        }
        else {
            return [NSString stringWithFormat:@"%@ Call (%@)",channel, status];
        }
    }
    else if (message.type == Bit6MessageType_Location) {
        return @"Location";
    }
    else if (message.type == Bit6MessageType_Attachments) {
        return @"Attachment";
    }
    else {
        return message.content;
    }
}

@end

File: CoreSamples/Bit6FullDemo-Swift/Bit6ChatDemo-Swift/ChatsTableViewController.swift
//
//  ChatsTableViewController.swift
//  Bit6ChatDemo-Swift
//
//  Created by Carlos Thurber Boaventura on 07/08/14.
//  Copyright (c) 2014 Bit6. All rights reserved.
//

import UIKit
import MobileCoreServices
import Bit6

class ChatsTableViewController: UITableViewController, UITextFieldDelegate {

    var conversation : Bit6Conversation! {
        didSet {
            Bit6.setCurrentConversation(conversation)
            NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(ChatsTableViewController.messagesChangedNotification(_:)), name: Bit6MessagesChangedNotification, object: self.conversation)
            NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(ChatsTableViewController.groupsChangedNotification(_:)), name: Bit6GroupsChangedNotification, object: nil)
            
            NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(ChatsTableViewController.typingBeginNotification(_:)), name: Bit6TypingDidBeginRtNotification, object: nil)
            NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(ChatsTableViewController.typingEndNotification(_:)), name: Bit6TypingDidEndRtNotification, object: nil)
        }
    }
    
    var scroll : Bool
    var _messages : [Bit6Message]!
    var messages : [Bit6Message] {
        get {
            if _messages != nil{
                return _messages
            }
            else {
                _messages = self.conversation.messages
                return _messages
            }
        }
        set {
            _messages = newValue
        }
    }
    
    
    @IBOutlet var typingBarButtonItem: UIBarButtonItem!
    
    required init?(coder aDecoder: NSCoder) {
        self.scroll = false
        super.init(coder:aDecoder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let userIdentity = Bit6.session().activeIdentity {
            self.navigationItem.prompt = "Logged as \(userIdentity.uri)"
        }
    
        let callItem = UIBarButtonItem(image:UIImage(named:"bit6ui_phone"), style:.Plain, target:self, action:#selector(ChatsTableViewController.call))
        
        self.navigationItem.rightBarButtonItem = callItem
        
        if let typingAddress = self.conversation.typingAddress {
            self.typingBarButtonItem.title = "\(typingAddress.uri) is typing..."
        }
        else {
            self.typingBarButtonItem.title = nil;
        }
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        self.navigationController?.setToolbarHidden(false, animated: true)
        if !self.scroll {
            self.scroll = true
            self.tableView.reloadData()
            self.tableView.setContentOffset(CGPointMake(0, CGFloat.max), animated: false)
        }
    }
    
    deinit {
        Bit6.setCurrentConversation(nil)
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK: - TableView

    override func tableView(tableView: UITableView?, numberOfRowsInSection section: Int) -> Int {
        return self.messages.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {

        let message = self.messages[indexPath.row]
        var cell : UITableViewCell
        
        if message.incoming {
            cell = tableView.dequeueReusableCellWithIdentifier("textInCell")!
        }
        else {
            cell = tableView.dequeueReusableCellWithIdentifier("textOutCell")!
        }
        
        return cell
    }
    
    override func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
        let message = self.messages[indexPath.row]
        
        if let textLabel = cell.viewWithTag(1) as? UILabel {
            textLabel.text = ChatsTableViewController.textContentForMessage(message)
        }
        
        if let detailTextLabel = cell.viewWithTag(2) as? UILabel {
            if message.type == .Call {
                detailTextLabel.text = ""
            }
            else {
                switch message.status {
                case .New : detailTextLabel.text = ""
                case .Sending : detailTextLabel.text = "Sending"
                case .Sent : detailTextLabel.text = "Sent"
                case .Failed : detailTextLabel.text = "Failed"
                case .Delivered : detailTextLabel.text = "Delivered"
                case .Read : detailTextLabel.text = "Read"
                }
            }
        }
    }
    
    // MARK: - Calls
    
    //WARNING: change to Bit6CallMediaModeMix to enable recording during a call
    static let callMediaMode = Bit6CallMediaModeP2P
    
    func call() {
        let actionSheet = UIAlertController(title:nil, message: nil, preferredStyle: .ActionSheet)
        
        actionSheet.addAction(UIAlertAction(title: "Audio Call", style: .Default, handler:{(action :UIAlertAction) in
            Bit6.startCallTo(self.conversation.address, streams:[.Audio], mediaMode:ChatsTableViewController.callMediaMode, offnet: false)
        }))
        actionSheet.addAction(UIAlertAction(title: "Video Call", style: .Default, handler:{(action :UIAlertAction) in
            Bit6.startCallTo(self.conversation.address, streams:[.Audio,.Video], mediaMode:ChatsTableViewController.callMediaMode, offnet: false)
        }))
        
        actionSheet.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler:nil))
        
        self.navigationController?.presentViewController(actionSheet, animated: true, completion:nil)
    }
    
    // MARK: - Send Text
    
    @IBAction func touchedComposeButton(sender : UIBarButtonItem) {
        
        if !self.canChat() {
            let alert = UIAlertController(title:"You have left this group", message: nil, preferredStyle: .Alert)
            alert.addAction(UIAlertAction(title: "OK", style: .Default, handler:{(action :UIAlertAction) in
            }))
            self.navigationController?.presentViewController(alert, animated: true, completion:nil)
            return;
        }
        
        let alert = UIAlertController(title:"Type the message", message: nil, preferredStyle: .Alert)
        
        alert.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler:{(action :UIAlertAction) in
            
        }))
        alert.addAction(UIAlertAction(title: "Done", style: .Default, handler:{(action :UIAlertAction) in
            let msgTextField = alert.textFields![0]
            
            guard let text = msgTextField.text else {
                return
            }
            
            self.sendTextMsg(text)
        }))
        
        alert.addTextFieldWithConfigurationHandler({(textField: UITextField) in
            textField.placeholder = "Message"
            textField.delegate = self
        })
        
        self.navigationController?.presentViewController(alert, animated: true, completion:nil)
    }

    func sendTextMsg(msg : NSString){
        if msg.length > 0 {
            let message = Bit6OutgoingMessage(destination:self.conversation.address)
            message.content = msg as String
            message.sendWithCompletionHandler{ (response, error) in
                if error == nil {
                    NSLog("Message Sent")
                }
                else {
                    NSLog("Message Failed with Error: \(error?.localizedDescription)")
                }
            }
        }
    }
    
    // MARK: - Typing
    
    func typingBeginNotification(notification:NSNotification) {
        let userInfo = notification.userInfo!
        let fromAddress = userInfo[Bit6FromKey] as! Bit6Address
        let convervationAddress = notification.object as! Bit6Address
        if convervationAddress == self.conversation.address {
            self.typingBarButtonItem.title = "\(fromAddress.uri) is typing..."
        }
    }
    
    func typingEndNotification(notification:NSNotification) {
        let convervationAddress = notification.object as! Bit6Address
        if convervationAddress == self.conversation.address {
            self.typingBarButtonItem.title = nil
        }
    }
    
    // MARK: - Data Source changes
    
    //here we listen to changes in group members and group title
    func groupsChangedNotification(notification:NSNotification) {
        let userInfo = notification.userInfo!
        let object = userInfo[Bit6ObjectKey] as! Bit6Group
        let change = userInfo[Bit6ChangeKey] as! String
        
        if change == Bit6UpdatedKey && object.address == self.conversation.address {
            self.title = ConversationsViewController.titleForConversation(self.conversation)
        }
    }
    
    func messagesChangedNotification(notification:NSNotification) {
        var userInfo = notification.userInfo!
        let object = userInfo[Bit6ObjectKey] as! Bit6Message
        let change = userInfo[Bit6ChangeKey] as! String
        
        if change == Bit6AddedKey {
            let indexPath = NSIndexPath(forRow: self.messages.count, inSection: 0)
            self.messages.append(object)
            self.tableView.insertRowsAtIndexPaths([indexPath], withRowAnimation:.Automatic)
            self.scrollToBottomAnimated(true)
        }
        else if change == Bit6UpdatedKey {
            if let indexPath = findMessage(object) {
                let cell = self.tableView.cellForRowAtIndexPath(indexPath)
                if cell != nil {
                    self.tableView(self.tableView, willDisplayCell:cell!, forRowAtIndexPath:indexPath)
                }
            }
        }
        else if change == Bit6DeletedKey {
            if let indexPath = findMessage(object) {
                self.messages.removeAtIndex(indexPath.row)
                self.tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation:.Automatic)
            }
        }
    }
    
    func findMessage(message:Bit6Message) -> NSIndexPath? {
            for x in (0..<self.messages.count).reverse() {
            if self.messages[x].isEqual(message) {
                return NSIndexPath(forRow:x, inSection:0)
            }
        }
        return nil
    }
    
    // MARK: -
    
    func scrollToBottomAnimated(animated:Bool){
        if self.messages.count>0 {
            let section = 0
            let row = self.tableView(self.tableView, numberOfRowsInSection: section)-1
            let scrollIndexPath = NSIndexPath(forRow: row, inSection: section)
            self.tableView.scrollToRowAtIndexPath(scrollIndexPath, atScrollPosition: UITableViewScrollPosition.Bottom, animated: animated)
        }
    }
    
    // MARK: - UITextFieldDelegate
    
    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool
    {
        Bit6.typingBeginToAddress(self.conversation.address)
        return true
    }
    
    // MARK: - Helpers
    
    func canChat() -> Bool {
        if let group = Bit6Group(conversation:self.conversation) {
            if group.hasLeft {
                return false
            }
        }
        return true
    }
    
    class func textContentForMessage(message:Bit6Message) -> String? {
        if message.type == .Call {
            var showDuration = false
            
            var status = ""
            switch message.callStatus {
            case .Answer: status = "Answer"; showDuration = true;
            case .Missed: status = "Missed";
            case .Failed: status = "Failed";
            case .NoAnswer: status = "No Answer";
            }
            
            var channels = [String]()
            
            if message.callHasChannel(.Audio) {
                channels.append("Audio")
            }
            if message.callHasChannel(.Video) {
                channels.append("Video");
            }
            if message.callHasChannel(.Data) {
                channels.append("Data");
            }
            
            let channel = channels.joinWithSeparator(" + ")
            
            if showDuration {
                return "\(channel) Call - \(message.callDuration!.description)s"
            }
            else {
                return "\(channel) Call (\(status))"
            }
        }
        else if message.type == .Location {
            return "Location"
        }
        else if message.type == .Attachments {
            return "Attachment"
        }
        else {
            return message.content;
        }
    }
}
